File: src/Java/myservicebus/src/main/java/com/myservicebus/JobConsumerRegistry.java
package com.myservicebus;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Semaphore;

import com.myservicebus.di.ServiceProvider;

public final class JobConsumerRegistry {
    public static final class Descriptor {
        private final Class<?> consumerType;
        private final Class<?> jobType;
        private final JobConsumerOptions options;
        private final String jobTypeName;
        private final Semaphore concurrency;

        Descriptor(Class<?> consumerType, Class<?> jobType, JobConsumerOptions options) {
            this.consumerType = consumerType;
            this.jobType = jobType;
            this.options = options;
            jobTypeName = options.getJobTypeName() != null ? options.getJobTypeName() : jobType.getSimpleName();
            concurrency = new Semaphore(options.getConcurrentJobLimit());
        }

        public Class<?> jobType() {
            return jobType;
        }

        public JobConsumerOptions options() {
            return options;
        }

        public String jobTypeName() {
            return jobTypeName;
        }

        public Semaphore concurrency() {
            return concurrency;
        }

        @SuppressWarnings("unchecked")
        public java.util.concurrent.CompletionStage<Void> run(ServiceProvider services, JobExecutionContext context)
                throws Exception {
            JobConsumer<Object> consumer = (JobConsumer<Object>) services.getRequiredService(consumerType);
            return consumer.run(new InMemoryJobContext<>(context, context.job()));
        }
    }

    private final Map<Class<?>, Descriptor> descriptors = new ConcurrentHashMap<>();
    private final Map<String, Descriptor> descriptorsByName = new ConcurrentHashMap<>();

    void add(Class<?> consumerType, Class<?> jobType, JobConsumerOptions options) {
        Descriptor descriptor = new Descriptor(consumerType, jobType, options);
        if (descriptors.containsKey(jobType)) {
            throw new IllegalStateException("A job consumer is already registered for " + jobType.getName());
        }
        if (descriptorsByName.containsKey(descriptor.jobTypeName())) {
            throw new IllegalStateException(
                    "A job consumer is already registered for job type name '" + descriptor.jobTypeName() + "'");
        }

        descriptors.put(jobType, descriptor);
        descriptorsByName.put(descriptor.jobTypeName(), descriptor);
    }

    public Descriptor get(Class<?> jobType) {
        Descriptor descriptor = descriptors.get(jobType);
        if (descriptor == null) {
            throw new IllegalStateException("No job consumer is registered for " + jobType.getName());
        }
        return descriptor;
    }

    public Descriptor get(String jobTypeName) {
        Descriptor descriptor = descriptorsByName.get(jobTypeName);
        if (descriptor == null) {
            throw new IllegalStateException("No job consumer is registered for job type name '" + jobTypeName + "'");
        }
        return descriptor;
    }
}

File: src/MyServiceBus/JobConsumerRegistry.cs
using Microsoft.Extensions.DependencyInjection;

namespace MyServiceBus;

public interface IRegisteredJobConsumer
{
    Type JobType { get; }

    string JobTypeName { get; }

    JobConsumerOptions Options { get; }

    SemaphoreSlim Concurrency { get; }

    Task Run(IServiceProvider serviceProvider, JobExecutionContext context);
}

internal sealed class JobConsumerDescriptor<TConsumer, TJob> : IRegisteredJobConsumer
    where TConsumer : class, IJobConsumer<TJob>
    where TJob : class
{
    public JobConsumerDescriptor(JobConsumerOptions options)
    {
        Options = options;
        JobTypeName = options.JobTypeName ?? typeof(TJob).Name;
        Concurrency = new SemaphoreSlim(options.ConcurrentJobLimit, options.ConcurrentJobLimit);
    }

    public Type JobType => typeof(TJob);

    public string JobTypeName { get; }

    public JobConsumerOptions Options { get; }

    public SemaphoreSlim Concurrency { get; }

    public Task Run(IServiceProvider serviceProvider, JobExecutionContext context)
    {
        var consumer = serviceProvider.GetRequiredService<TConsumer>();
        return consumer.Run(new InMemoryJobContext<TJob>(context, (TJob)context.Job));
    }
}

public interface IJobConsumerRegistry
{
    IRegisteredJobConsumer Get(Type jobType);

    IRegisteredJobConsumer Get(string jobTypeName);
}

internal sealed class JobConsumerRegistry : IJobConsumerRegistry
{
    private readonly Dictionary<Type, IRegisteredJobConsumer> descriptors = [];
    private readonly Dictionary<string, IRegisteredJobConsumer> descriptorsByName =
        new(StringComparer.Ordinal);

    public void Add<TConsumer, TJob>(JobConsumerOptions options)
        where TConsumer : class, IJobConsumer<TJob>
        where TJob : class
    {
        var descriptor = new JobConsumerDescriptor<TConsumer, TJob>(options);
        if (descriptors.ContainsKey(typeof(TJob)))
            throw new InvalidOperationException($"A job consumer is already registered for {typeof(TJob)}.");
        if (descriptorsByName.ContainsKey(descriptor.JobTypeName))
            throw new InvalidOperationException($"A job consumer is already registered for job type name '{descriptor.JobTypeName}'.");

        descriptors.Add(typeof(TJob), descriptor);
        descriptorsByName.Add(descriptor.JobTypeName, descriptor);
    }

    public IRegisteredJobConsumer Get(Type jobType) =>
        descriptors.TryGetValue(jobType, out var descriptor)
            ? descriptor
            : throw new InvalidOperationException($"No job consumer is registered for {jobType}.");

    public IRegisteredJobConsumer Get(string jobTypeName) =>
        descriptorsByName.TryGetValue(jobTypeName, out var descriptor)
            ? descriptor
            : throw new InvalidOperationException($"No job consumer is registered for job type name '{jobTypeName}'.");
}
